### PR TITLE
Issue #718 Downloading initial ISO and oc binary should support HTTP proxy

### PR DESCRIPTION
**Pull request by praveenkumar**
This patch will modify `http` client in case of proxy and use it to download iso and oc binary. Please test it, for me working as expected I can see traffic is going from proxy server.

### CODE REVIEW

**Comment by praveenkumar** [comment]
> Why do all CI fail?

I do need to update unit test also, will do that and add as [WIP] for it.

File: cmd/minishift/cmd/start.go [hunk 1]
```diff
@@ -347,9 +349,11 @@ func setOcProxy() {
 func setShellProxy() {
 	if viper.IsSet(httpProxy) {
 		shellProxyEnv += fmt.Sprintf("http_proxy=%s ", viper.GetString(httpProxy))
+		proxyUrl = viper.GetString(httpProxy)
 	}
 	if viper.IsSet(httpsProxy) {
 		shellProxyEnv += fmt.Sprintf("https_proxy=%s ", viper.GetString(httpsProxy))
+		proxyUrl = viper.GetString(httpProxy)
```
Comment (LalatenduMohanty) [inline cmd/minishift/cmd/start.go]: should be viper.GetString(httpsProxy)

File: pkg/minikube/cluster/cluster.go [hunk 2]
```diff
@@ -214,7 +215,13 @@ func (m *MachineConfig) CacheMinikubeISOFromURL() error {
 	fmt.Println(fmt.Sprintf("Downloading ISO '%s'", m.MinikubeISO))
 
 	// store the iso inside the MINISHIFT_HOME dir
-	response, err := http.Get(m.MinikubeISO)
+	myClient := http.DefaultClient
```
Comment (LalatenduMohanty) [inline pkg/minikube/cluster/cluster.go]: 👍

Comment (gbraad) [inline pkg/minikube/cluster/cluster.go]: Code duplication. This is a candidate for a general download function that can be part of our utility library

```
Get(uri string, proxy ProxyInfo) error {}
```

File: pkg/util/github/github.go [hunk 3]
```diff
@@ -120,13 +121,18 @@ func DownloadOpenShiftReleaseBinary(binaryType OpenShiftBinaryType, osType minis
...
-	asset, url, err := client.Repositories.DownloadReleaseAsset("openshift", "origin", assetID)
+	asset, uri, err := client.Repositories.DownloadReleaseAsset("openshift", "origin", assetID)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("Cannot download OpenShift release asset %d", assetID))
 	}
-	if len(url) > 0 {
+	if len(uri) > 0 {
 		fmt.Println(fmt.Sprintf("Downloading OpenShift binary '%s' version '%s'", binaryType.String(), *release.TagName))
-		httpResp, err := http.Get(url)
+		myClient := http.DefaultClient
```
Comment (gbraad) [inline pkg/util/github/github.go]: See above; code duplication

**Comment by gbraad** [comment]
BTW, tested on Windows with Polipo as a caching server. Didn't want to dust off and test my Squid setup skills... but worked as expected.

File: pkg/util/github/github.go [hunk 4]
```diff
@@ -120,13 +121,14 @@ func DownloadOpenShiftReleaseBinary(binaryType OpenShiftBinaryType, osType minis
...
-	asset, url, err := client.Repositories.DownloadReleaseAsset("openshift", "origin", assetID)
+	asset, uri, err := client.Repositories.DownloadReleaseAsset("openshift", "origin", assetID)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("Cannot download OpenShift release asset %d", assetID))
 	}
-	if len(url) > 0 {
+	if len(uri) > 0 {
 		fmt.Println(fmt.Sprintf("Downloading OpenShift binary '%s' version '%s'", binaryType.String(), *release.TagName))
-		httpResp, err := http.Get(url)
+		myClient := util.GetHttpClient(proxyUrl)
```
Comment (gbraad) [inline pkg/util/github/github.go]: Ah, I see what you did.
For some reason, I do think this code is more complex than it needs to be. Readability is not very nice...

**Comment by LalatenduMohanty** [comment]
Changes looks good to me except that the openshift_caching.go is not used in the code. Had a discussion with @praveenkumar and he will create a followup issue for removing it. cc @hferentschik

**Comment by praveenkumar** [comment]
@LalatenduMohanty @hferentschik #937 Issue created.

**Comment by hferentschik** [comment]
@praveenkumar have you tried to configure the default transport? See also https://stackoverflow.com/questions/14661511/setting-up-proxy-for-http-client. This is probably even better than env variables. In fact this is what I had in mind first. If the proxy variables are set, we configure the default transport, so while the Minisfhit process is running, all requested HTTP clients (unless explicitly configured) use the default transport. This will then also work with the upcoming _update_ command which will also need to support proxies.

**Comment by hferentschik** [comment]
More concretely, I would expect that we make this call when proxies are specified via the CLI

    http.DefaultTransport = &http.Transport{Proxy: http.ProxyURL(proxyUrl)}

**Comment by praveenkumar** [comment]
@hferentschik Changing default transport is working (I did test it out) now how we should go about to add it. Should we revert committed code and then again send PR?

**Comment by hferentschik** [comment]
> Changing default transport is working

There is a problem with it so, in terms how it deals with HTTP vs HTTPS proxies. I think the env approach is better, since it transparently makes the right decision on which proxy to use. 

 > (I did test it out) now how we should go about to add it. Should we revert committed code and then again send PR?

I think the best is to add this commit https://github.com/minishift/minishift/pull/942/commits/d046e2e24971dbca035f5742626dc2fe30b9811c which inverts the changes and either add another commit with the new implementation or even squash in a single commit.

**Pull request merged**